src/tools/tidy/src/rustdoc_html_ids.rs

```diff
@@ -0,0 +1,164 @@
+//! Checks that the rustdoc ID map is up-to-date. The goal here is to check a few things:
+//!
+//! * All IDs created by rustdoc (through JS or files generation) are declared in the ID map.
+//! * There are no unused IDs.
+
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::Path;
+
+use regex::Regex;
+
+const ID_MAP_PATH: &str = "librustdoc/html/markdown.rs";
+const IDS_USED_IN_JS: &[&str] = &[
+    // This one is created in the JS and therefore cannot be found in rust files.
+    "help",
+];
+
+fn extract_ids(path: &Path, bad: &mut bool) -> HashMap<String, usize> {
+    let file = File::open(path).expect("failed to open file to extract rustdoc IDs");
+    let buf_reader = BufReader::new(file);
+    let mut iter = buf_reader.lines();
+    let mut ids = HashMap::new();
+
+    while let Some(Ok(line)) = iter.next() {
+        if line.trim_start().starts_with("html_id_map!(") {
+            break;
+        }
+    }
+    // We're now in the function body, time to retrieve the IDs!
+    while let Some(Ok(line)) = iter.next() {
+        let line = line.trim_start();
+        if line.starts_with("// ") {
+            // It's a comment, ignoring this line...
+            continue;
+        } else if line.starts_with(")") {
+            // We reached the end of the IDs declaration list.
+            break;
+        }
+        let id = line.split('"').skip(1).next().unwrap();
+        if ids.insert(id.to_owned(), 0).is_some() {
+            eprintln!(
+                "=> ID `{}` is defined more than once in the ID map in file `{}`",
+                id, ID_MAP_PATH
+            );
+            *bad = true;
+        }
+    }
+    if ids.is_empty() {
+        eprintln!("=> No IDs were found in rustdoc in file `{}`...", ID_MAP_PATH);
```

Comment:
We actually scanned `path`, not `ID_MAP_PATH`:

```suggestion
        eprintln!("=> No IDs were found in rustdoc in file `{}`...", path);
```

Reply:
They're the same but I'll replace it here (and in the `eprintln!` above as well) to make it more logical.